tt_metal/impl/dispatch/command_queue.cpp

```diff
@@ -828,99 +828,96 @@ void EnqueueProgramCommand::assemble_device_commands(
-                        kernel_bins_unicast_cmds.emplace_back(2 * CQ_PREFETCH_CMD_BARE_MIN_SIZE);
-                        cmd_sequence_sizeB += 2 * CQ_PREFETCH_CMD_BARE_MIN_SIZE;
-                        kernel_bins_unicast_cmds.back().add_dispatch_write_linear(
-                            false,                // flush_prefetch
-                            dst_noc_info.second,  // num_mcast_dests
-                            noc_encoding,         // noc_xy_addr
-                            kg_transfer_info.dst_base_addrs[kernel_idx],
-                            kg_transfer_info.lengths[kernel_idx]);
-                        // Difference between prefetch total relayed pages and dispatch write linear
-                        uint32_t relayed_bytes =
-                            align(kg_transfer_info.lengths[kernel_idx], HostMemDeviceCommand::PROGRAM_PAGE_SIZE);
-                        // length_adjust needs to be aligned to NOC_DRAM_ALIGNMENT
-                        uint16_t length_adjust = uint16_t(relayed_bytes - kg_transfer_info.lengths[kernel_idx]);
-
-                        uint32_t base_address, page_offset;
-                        if (kg_transfer_info.page_offsets[kernel_idx] > CQ_PREFETCH_RELAY_PAGED_START_PAGE_MASK) {
-                            const uint32_t num_banks =
-                                this->device->num_banks(this->program.kg_buffers[buffer_idx]->buffer_type());
-                            page_offset = kg_transfer_info.page_offsets[kernel_idx] % num_banks;
-                            uint32_t num_full_pages_written_per_bank =
-                                kg_transfer_info.page_offsets[kernel_idx] / num_banks;
-                            base_address =
-                                this->program.kg_buffers[buffer_idx]->address() +
-                                num_full_pages_written_per_bank * this->program.kg_buffers[buffer_idx]->page_size();
-                        } else {
-                            base_address = this->program.kg_buffers[buffer_idx]->address();
-                            page_offset = kg_transfer_info.page_offsets[kernel_idx];
-                        }
-
-                        kernel_bins_unicast_cmds.back().add_prefetch_relay_paged(
-                            true,  // is_dram
-                            page_offset,
-                            base_address,
-                            this->program.kg_buffers[buffer_idx]->page_size(),
-                            relayed_bytes / this->program.kg_buffers[buffer_idx]->page_size(),
-                            length_adjust);
+        for (const auto& [cores, num_mcast_dests, kg_transfer_info] : program.program_transfer_info.kernel_bins) {
+            bool write_linear;
+            uint32_t noc_encoding;
+            std::visit(
+                [&](auto&& cores) {
+                    using T = std::decay_t<decltype(cores)>;
+                    if constexpr (std::is_same_v<T, CoreRange>) {
```

Comment:
Is this generic for WORKER and ETH? What if the EthernetKernel uses `CoreRange((0,0) to (0,2))`? I don't think that's asserted anywhere. It will cause this check to bug out for ETH, maybe add to the CreateEthernetKernel an assert that cores have to be CoreCoord for WH? In BH it's probably ok, depending on how `get_noc_multicast_encoding` gets uplifted.

Reply:
This is generic for workers and eth and is how main is currently set up. The code for this part hasn't changed.

This is not affected by how the user creates their kernel as this comes from how we encoded the coords in `tt_metal/impl/program/program.cpp`.

```
for (KernelGroup &kernel_group : this->get_kernel_groups(CoreType::WORKER)) {
         std::vector<pair<transfer_info_cores, uint32_t>> dst_noc_multicast_info = extract_dst_noc_multicast_info<std::set<CoreRange>>(
             device, kernel_group.core_ranges.ranges(), kernel_group.get_core_type());
```
The code above generates the CoreRanges consumed by the code you highlighted, whereas the code below generates the CoreCoords. command_queue.cpp is agnostic to whether it is eth or worker.
```
for (KernelGroup &kernel_group : this->get_kernel_groups(CoreType::ETH)) {
        vector<pair<transfer_info_cores, uint32_t>> dst_noc_unicast_info =
            extract_dst_noc_unicast_info(kernel_group.core_ranges.ranges(), kernel_group.get_core_type());
```

Reply:
Thanks, I forgot my own work 🐒